Fail on unknown gw

DunderRoffe opened the pull request.

Commit:
GatewayConfig: Add constant ids member function

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
SoftwareContainer: assert valid state in confGW

Call asserValidState in the configureGateway method

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
SoftwareContainer: Check gwconf map to real gws

Throw an exception when finding ids in a given GatewayConfiguration
that do not map to any Gateway in SoftwareContainer

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

tobsan reviewed (approved):
This looks good. One could change the m_gateways member though to be a map instead of a vector, with IDs as keys.

DunderRoffe commented:
Yes that is a good idea. However, I fell that it is out of scope for this pull-request.

kursatkobya commented on libsoftwarecontainer/src/softwarecontainer.cpp:
```diff
@@ -197,6 +197,19 @@ ReturnCode SoftwareContainer::startGateways(const GatewayConfiguration &gwConfig
+    // to a gateway in SoftwareContainer
+    for (auto &id : gwConfig.ids()) {
+        for (auto &gateway : m_gateways) {
+            if (id != m_gateways) {
```
how come you compare string id with vector of  Gateway pointer
is this really compiles ? Or maybe i misread the code

Commit:
GatewayConfig: Add janson independent append

Add an overloaded append function that takes a std::string instead of a
janson JSON pointer. This ease the creation of component tests and allows
for creation of launchers that do not use the janson library.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
SoftwareContainer: Test for non existing gw ids

Add a component test to test that supplying a GatewayConfig to SoftwareContainer
containing an gateway-id that does not map to a loaded Gateway fails.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
GatewayError: Make a subclass of SC-Error

Make GatewayError a subclass of SoftwareContainerError. In order to be able
to do proper error handling in launchers implementing the library, all
exceptions thrown by SoftwareContainer should be subclasses of
SoftwareContainerError.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
ServiceTests: Remove xfail for capabilites test

Remove xfail flag from test_setting_capability_with_unkown_gw_id_fails
as it now works properly.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

Commit:
ServiceTest: Add regression test for 2x setNetGW

Add regression test that tries to set the network gateway twice. This should work
without the agent crashing due to glibmm error.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

kursatkobya commented on common/gatewayconfig.cpp:
```diff
@@ -39,6 +39,22 @@ GatewayConfiguration::~GatewayConfiguration()
+    size_t flags = 0; // default flags
+
+    json_t *json = json_loads(jsonConf.c_str(), flags, &jsonError);
+    if (json == nullptr) {
```
it would be nice if you write other way around : `nullptr == json`

kursatkobya commented on libsoftwarecontainer/src/softwarecontainer.cpp:
```diff
@@ -197,6 +197,23 @@ ReturnCode SoftwareContainer::startGateways(const GatewayConfiguration &gwConfig
+            }
+        }
+        if (!match) {
+            throw GatewayError("Could not find any gateway matching id: " + id);
```
Maybe this is a really small possibility but if there is no enabled gateway this could create false throw. Maybe it would be reasonable to check if there is any member in `m_gateways`.

DunderRoffe replied:
This is how it should be. Specifying config for non-existent/loaded gateways is not supported.

kursatkobya reviewed (approved):
Looks nice. I gave small minor comments with my approval.
